### ISSUE TITLE
Add support for signed tokens in proteus-[events, frontend]

### ISSUE THREAD

**Issue by hellais** [ooni/orchestra#27]
What this means is that:

* [ ] When a new orchestration request is created a to be signed token needs to be created and output in the web UI
* [ ] The signed tokens need to be stored in the database and sent to clients out there

Something tricky that may appear with doing this is that currently the orchestration requests are unique per target user. It's not an issue given the current design for it to work like this, because at the moment every client is anyways receiving the same orchestration request, however when we have smarter scheduling algorithms, this will become an issue as if each request sent to clients needs to be different, then we need to sign every request separately.

The extra challenge is that signing on HSM is quite slow, yubikeys claim to be able to make ~1 signature per second, so if we have 10'000 clients, this means it's going to take almost 3 hours to sign all the requests. This is clearly not acceptable.

A possible solution is that what operators sign are actually just some sort of general rule and the clients then compare if the unit of work they receive matches the constraints defined in the general global rule.

An alternative way of doing it is that all the possible requests for clients are wrapped inside of one unique object that is all signed. The client will then read this huge blob and just take out the unit of work that is relevant to them.

Each approach has it's tradeoffs.

To illustrate this problem I think it's useful to make an example:

Say an operator goes to the orchestration interface and defines the following orchestration request (note the schema and format is not formal or how it will really look inside of the DB to make the example simpler to understand):

"All probes (`country_code: *`) should test the websites (`test_name: web_connectivity`) inside of the citizenlab testing list category for "Search Engines" (`category_code: SRCH`)"

This gets transformed by into the following **rule**:

```
{
"country_code": "*",
"test_name": "web_connectivity",
"category_code": "SRCH"
}
```

In turn this rule generates the following **orchestration request** to be sent to clients:

```
{
"country_code": "*",
"test_name": "web_connectivity",
"urls": [
"https://bing.com",
"https://google.com",
"https://yandex.ru"
]
}
```

Now with the naive scheduling algorithm, this is not too much of a problem as the same request will be issued to every client out there. So we can just sign the above **orchestration request**. However if we wanted to do something more complex, where say some number of clients will test "https://bing.com", while others test "https://google.com", we would have to sign each individual request separately.

With 3 websites it's not too much work, but you can imagine this starting to get a bit out of hand when you have thousands of websites and you need to make combinations of these sites on a per probe basis.

**Comment by hellais** [comment]
So I think I have an idea for solving the issue of having different per client tasks, but having secure signing.

The issue is mainly a problem when dealing with tests that have orchestration arguments. In general we can abstract the problem as follows:

The operators defines some task that will be spread out over all the clients. The task defined by the operator in fact is divided into a set of N units. To exemplify this, when I say "all probes in Italy" (the target) "run web_connectivity" (the test name) "on the global list" (the tasks), the tasks are actually a list, in this case a list of URLs.

What we could do is actually sign the above section and then inside of the orchestration request encode a sort of pointer into the task list.

For example the following request (assuming what is in the `urls` is the content of global test list):

```
{
"country_code": "IT",
"test_name": "web_connectivity",
"args": [
{"url": "https://bing.com"},
{"url": "https://google.com"},
{"url": "https://yandex.ru"}
]
}
```

Would be the thing signed by the operator. While probes that are to run a test only on `https://bing.com` and `https://yandex.ru` will receive something like:

```
{
  "signed_task": JWS({"country_code": "IT","test_name": "web_connectivity","args":[{"url": "https://bing.com"},{"url": "https://google.com"},{"url": "https://yandex.ru"}]}
  "args_idx": [0,1]
}
```

Since `urls_idx` is not actually signed, we can generate that dynamically in the orchestration backend.

The tradeoff of this approach is that somebody without a signature can create "non approved" combinations of task arguments. I don't see a way in which this can be an issue (if the full list of URLs is safe, so should a subset of them).

When we add new sites to the country (or global testing lists) we will still have to resign stuff, but we can do that with some contextual alerts that appear to the operators that need to update their signatures and it should not be a huge burden.

**Comment by bassosimone** [comment]
@hellais so, you're leaning towards the approach where the "big" request is signed and then the client is instructed to only consider a subset of it using unsigned information?

**Comment by hellais** [comment]
> @hellais so, you're leaning towards the approach where the "big" request is signed and then the client is instructed to only consider a subset of it using unsigned information?

yes, correct. I think by keeping the size of the instruction that points to the subset of information and making it just a pointer into the signed data, we can actually be quite safe (you don't have to implement set intersection and constraint satisfaction logic into the client code) and not use up too much space.

**Comment by bassosimone** [comment]
Okay. I spent some more time thinking about it, and it seems to me the solution you propose is good for the use case where we have a somewhat large unit of work and want to spread the load among clients.

The only concern that I have is that this may become inefficient in specific scenarios (especially for mobile and, even more especially, for countries where the connectivity is bad).

I am also a bit concerned that I do not fully understand how it would work, so I am wrongly estimating the impact of the proposed change. My working assumption is that, given the rule in the example, user _Johnny_ will receive an orchestration response (including the full list, signed, and the differing indexes), for each time it needs to test some specific URLs. Is my understanding correct?

_Assuming my understanding to be correct_, as said, I wonder about the extra overhead:

Sure, it may be small considering a single response, but we are also talking of many repeated responses where the same information is sent to clients again and again and again. How many responses are we considering per day? Would the total amount of overhead bytes be acceptable (and workable) in places with slow and/or bad internet connectivity? How many URLs indexes would typically be present into a single response, one or many?

In the event where the response is no, I _believe_ we can think of an alternative schema. We can _refer to_ an orchestration list by URL inside of the response, like this:

```JSON
{
  "signed_task": "https://test-lists.orchestration.ooni.io/180cb5b6dcc83e67797a14e50f4fa4d50d6a826d9b84bf2f222f9520dbb92c2b",
  "args_idx": [0,1]
}
```

This is assuming that the client will fetch the specific list only once, if it has not already cached it. After download it, it will _verify_ the list. And the it can continue to use the list as long as it's valid.

**Comment by hellais** [comment]
It could make sense to implement some caching, but just to put the size we are talking about into perspective. For an instruction that is going to schedule running the global testing list the size we are talking about is actually not that much:

```
>>> len(json.dumps({'urls': urls}))
31259
```
So that is < 32KB.

If you apply gzip compression to the response:

```
>>> len(zlib.compress(json.dumps({'urls': urls})))
8223
```

So 8KB.

Given the fact that clients running a web_connectivity test will most likely be fetching pages that are probably larger than 32KB, I don't think this overhead is actually that much and maybe we are prematurely optimising.

**Comment by bassosimone** [comment]
> If you apply gzip compression to the response:

FYI: MK does not support that (yet).

I am going to consider the 32 KB scenario then. What I don't get is: how many times per day would this JSON be fetched by clients as part of orchestration?

**Comment by hellais** [comment]
> What I don't get is: how many times per day would this JSON be fetched by clients as part of orchestration?

The number of times depends on what schedule is configured in the orchestration request. For tasks that run a daily schedule, they will be fetched daily. For tasks that run on more frequent schedules it will be less frequently.

Since our currently scheduling algorithm is going to be the most naive one (i.e. test everything that is scheduled by everybody), I would say in most cases it will be at most daily (since we don't want clients to be testing the full list more than daily).

**Comment by hellais** [comment]
xref: https://github.com/measurement-kit/measurement-kit/issues/984

**Comment by bassosimone** [comment]
Alright. If the download is to be performed daily, I do not think this is currently a problem. Apart from this issue, I do not see other issues in the scheme you have proposed.

**Comment by hellais** [comment]
To recap, the data that we are going to be signing and wrapping in a JSON Signed Token is:

*country_code*: defines country of the intended test (could this also be a list of multiple countries). Scheduler can use it for distributing tasks and client can enforce this check as well

*test_name*: defines the type of test. eg. “web connectivity”

*list of urls*: taken from CL test list. Should include `url` and `category_code`, so that scheduler and clients can enforce preferences. eg. prevent scanning porn sites for high risk users

Also makes use of the `exp`iration time of JWK to prevent that signed data can be reused for ever.

*test interval & scheduling information*: 

Things that will not be signed are:

*url_idx*: offset into which of the specified urls to test

**Comment by hellais** [comment]
We are going to be discontinuing orchestra in the future. We can close this issue as no longer relevant.
